src/qb4-packw/kr-neoni8mm.c.in

```diff
@@ -0,0 +1,192 @@
+// Copyright 2023 Google LLC
+//
+// This source code is licensed under the BSD-style license found in the
+// LICENSE file in the root directory of this source tree.
+
+$ABC = "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ"
+
+$assert NR > 1
+$assert KR > 1
+#include <assert.h>
+#include <stddef.h>
+#include <stdint.h>
+#include <math.h>
+#include <arm_neon.h>
+
+#include "src/xnnpack/packw.h"
+
+// convert a vector from packed nibbles to planar, and accumulate sum
+static XNN_INTRINSIC
+int8x16_t xnn_packed2planar(
+  int32x4_t *vacc,
+  const uint8x16_t v,
+  const uint8x16_t vmask,
+  const uint8x16_t veor_mask,
+  const int32x4_t neg_zp,
+  const uint8x16_t vones_zero_mask)
+{
+  const uint8x16_t vl = vshrq_n_u8(v, 4);  // isolate lower int 4
+  const uint8x16_t vh = vandq_u8(v, vmask);  // isolate upper int 4
+  *vacc = vreinterpretq_s32_u32(vmmlaq_u32(vreinterpretq_u32_s32(*vacc), vones_zero_mask, vl));
```

Comment:
too much to have this template just for these instructions? can we drop it and just have dot prod one? What is the perf delta for the same output tile size between dot and i8mm?

Reply:
so perhaps this is bad naming of the template, but I think there is a good amount of difference is also due to kr=8 vs kr=4. Specifically the rearranging of the planes and blocks before we read xnn_packed2planar. I take advantage of the fact different KR is also different instructions to have different templates.